test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java

```diff
@@ -229,4 +231,104 @@ public void testChar() {
             Asserts.assertEquals(cspecial[i], (char) Math.abs(cspecial[i]));
         }
     }
- }
+
+    @Run(test = {"testIntRange1", "testIntRange2", "testIntRange3", "testIntRange4"})
+    public void checkIntRanges(RunInfo info) {
+        for (int i : ispecial) {
+            checkIntRange(i);
+        }
+
+        for (int j = 0; j < 20; j++) {
+            int i = Generators.G.ints().next();
+            checkIntRange(i);
+        }
+    }
+
+    @DontCompile
+    public void checkIntRange(int i) {
+        Asserts.assertEquals(Math.abs((i & 7) - 4) > 4, testIntRange1(i));
+        Asserts.assertEquals(Math.abs((i & 7) - 4) < 0, testIntRange2(i));
+        Asserts.assertEquals(Math.abs(-((i & 7) + 2)) < 2, testIntRange3(i));
+        Asserts.assertEquals(Math.abs(-((i & 7) + 2)) > 9, testIntRange4(i));
+    }
+
+    @Run(test = {"testLongRange1", "testLongRange2", "testLongRange3", "testLongRange4"})
+    public void checkLongRanges(RunInfo info) {
+        for (long l : lspecial) {
+          checkLongRange(l);
+        }
+
+        for (int j = 0; j < 20; j++) {
+            long l = Generators.G.longs().next();
+            checkLongRange(l);
+        }
+    }
+
+    @DontCompile
+    public void checkLongRange(long l) {
+        Asserts.assertEquals(Math.abs((l & 7) - 4) > 4, testLongRange1(l));
+        Asserts.assertEquals(Math.abs((l & 7) - 4) < 0, testLongRange2(l));
+        Asserts.assertEquals(Math.abs(-((l & 7) + 2)) < 2, testLongRange3(l));
+        Asserts.assertEquals(Math.abs(-((l & 7) + 2)) > 9, testLongRange4(l));
+    }
+
+    // Int ranges
+
+    @Test
+    @IR(failOn = { IRNode.ABS_I })
+    public boolean testIntRange1(int in) {
+        // [-4, 3] => [0, 4]
+        return Math.abs((in & 7) - 4) > 4;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.ABS_I })
+    public boolean testIntRange2(int in) {
+        // [-4, 3] => [0, 4]
+        return Math.abs((in & 7) - 4) < 0;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.ABS_I })
+    public boolean testIntRange3(int in) {
+        // [-9, -2] => [2, 9]
+        return Math.abs(-((in & 7) + 2)) < 2;
```

Comment:
Not sure if this is in scope for this PR, but `abs(x)` should be idealized into `0 - x` if x <= 0. This seems to be missing at the moment.

Reply:
This is a good observation! I'll do this in a followup patch, to keep this one focused on just the Value() function.